3rdParty/iresearch/core/utils/object_pool.hpp

```diff
@@ -131,7 +132,7 @@ class concurrent_stack : private util::noncopyable {
         return nullptr;
       }
 
-      new_head.node = head.node->next;
+      new_head.node = head.node->next.load(std::memory_order_relaxed);
```

Comment:
`next` is set in `push`:
https://github.com/arangodb/arangodb/blob/215bf7cab476a7ea9d4250bbde86f40b9771e69f/3rdParty/iresearch/core/utils/object_pool.hpp#L147-L153
The problem is that nodes are _reused_. Suppose a thread is currently trying to pop a node, so it loaded the current `head`, but is interrupted before it can read `next`. Now some other thread is faster, pops the node and puts it on the free list. Then the node can be taken from the free list and reinserted into the stack (that's why we need ABA protection on the head). This push operation will overwrite the value of `next` in the reused node, and this write operation conflicts with the read in the first thread, resulting in a data race.

Reply:
Yes, I see, but `head` and `new_head` are local variables on stack. And once load from `_head` is finished we have a copy here and no other writes to `_head` could affect us. Indeed `next` value may became totally invalid but as we don't dereference it we are safe.

Reply:
True, we never dereference the value we read from `next`, and the subsequent CAS operation is doomed to fail anyway. However, that doesn't change the fact that we have two conflicting non-atomic operations (write + read) on `next` that are not ordered by a happens-before relation.

Reply:
Ah, I thing I`ve got it. This conflict could happen if we are popping the same `node` object we are pushing from the other thread. But that could not happen due to logic of stack usage. Even more  - if due to some circumstances this will happen  - we are still not using obtained value. And you are correct to notice that CAS will fail and protect us. So adding atomic here would be a pessimization. Could we put here some TSan suppress comment?

Reply:
> This conflict could happen if we are popping the same node object we are pushing from the other thread. But that could not happen due to logic of stack usage.

Actually it can (and it does) happen, because nodes are reused. That is also the reason why we need cmpxchg16 to protect against the ABA problem. Using atomic here is not really a pessimization since `memory_order_relaxed` is sufficient, so it will translate to the same instructions on virtually all architectures anyway. This is not just about TSan complaining, but this really is a data race as defined in the standard. Even though it is _most likely_ benign in almost all cases, strictly speaking it is still UB.

Reply:
For the sake of completeness here is a sequence that shows how the data race occurs:
  - suppose we have a stack a node A
  - thread 1 executes pop, so it reads `head` and receives a reference to A, but before it can read A's `next` pointer, the thread is interrupted
  - thread 2 also executes pop and removes A from the stack. The stack is now empty.
     -> A has now been completely removed from the stack, but thread 1 does not yet know that!
  - thread 2 now reuses A and pushes a new value to the stack. _This requires a write to A's `next` pointer._
  - thread 1 now continues and reads A's `next` -> this is a data race with the write in the previous step